test/pylib/coverage_utils.py

```diff
@@ -0,0 +1,1293 @@
+        return FileType.UNRECOGNIZED
+
+
+ELF_TYPES = [FileType.EXEC_BIN,FileType.EXEC_SO]
+PROFILED_ELF_TYPES = [FileType.PROFILED_EXEC_BIN, FileType.PROFILED_EXEC_SO]
+
+# Convenience functions
+
+def convert_to_path_list(paths : PathLikeOrPathLikeCollection) -> List[Path]:
+    """ A convenience function to normalize PathLikeOrPathLikeCollection into
+        a PathCollection
+
+    Args:
+        paths (PathLikeOrPathLikeCollection): The parameter to convert
+
+    Returns:
+        List[Path]: A path list
+    """
+    if paths is None:
+        return []
+    elif isinstance(paths, str) or not isinstance(paths, IterableType):
+        return [Path(paths)]
+    return [ Path(path) for path in paths]
+
+def concurrency_to_semaphore(concurrency : ConcurrencyParam) -> asyncio.Semaphore:
+    """A convenience function to convert a `ConcurrencyParam` to be used to limit
+    concurrency of execution.
+
+    Args:
+        concurrency (ConcurrencyParam): The concurrency to be converted to a semaphore.
+
+    Returns:
+        asyncio.Semaphore: A semaphore according to the following semantics:
+        1. If concurrence's actual type is int, so a semaphore with `concurrency` units.
+        2. If `concurrency` is a semaphore, it will be returned as is
+        3. In any other case, a semaphore with 1 unit will be returned (no concurrency allow)
+    """
+    if isinstance(concurrency,int):
+        return asyncio.Semaphore(concurrency)
+    elif isinstance(concurrency,asyncio.Semaphore):
+        return concurrency
+    else:
+        return asyncio.Semaphore(1)
+
+async def verify_file_type(f : PathLike, file_type: Union[FileType,Iterable[FileType]]) -> bool:
+    """Verify that a file is of certain type
+
+    Args:
+        f (PathLike): The path to the file whose type needs to be verified
+        file_type (Union[FileType,Iterable[FileType]]): A collection of filetypes or a single
+        file type for the file to be verified against.
+
+    Returns:
+        bool: True if the file type matches one of `file_type` False otherwise.
+    """
+    if not isinstance(file_type,IterableType):
+        file_type = [ file_type ]
+    file_type = list(file_type)
+    return await FileType.get_file_type(f) in file_type
+
+@traced_func
+async def gather_limited_concurrency(*args, semaphore : asyncio.Semaphore,logger : LoggerType = COVERAGE_TOOLS_LOGGER,**kwargs) -> Iterable[Any]:
+    """ A wrapper around asyncio.gather that limits concurrency for the submitted tasks.
+    This function takes parameters as asyncio.gather with an additional mandatory `semaphore` param that
+    is the semaphore that will limit the tasks concurrency.
+    Note: The function looks a bit convoluted, but it is actually a nuance that arises from the semaphore.
+    Since asyncio.gather, doesn't stop already submitted tasks, it can cause deadlock on the semaphore. So
+    we should make sure that in case of a failure the all tasks are terminated unconditionally.
```

Comment:
I am not sure it will work well with the `return_exceptions` option of  gather.
But I guess I could just try it out on some toy example first :slightly_smiling_face:

Reply:
As I suspected, the proposed implementation just throws an exception and cancel tasks even if we pass
`return_exceptions = True` to `gather`

Reply:
True, that one won't work with `return_exceptions`. But this one will:
```
async def gather_limited_concurrency(*args: Coroutine, semaphore: asyncio.Semaphore, **kwargs) -> list[Any]:
    async with asyncio.TaskGroup() as tg:
        tasks = [tg.create_task(with_semaphore(semaphore, x)) for x in args]
    return await asyncio.gather(*tasks, **kwargs)
```
You aren't using `return_exceptions` anywhere, though.

Reply:
Also, task group has different semantics: it doesn't just cancel the subtasks, but also waits for them to end. But I think that's a good thing.

Reply:
Anyway, yes, if we want `return_exceptions` to work, then the above doesn't work. You'd have to wrap each coroutine in a try..except.